metacat-main/src/main/java/com/netflix/metacat/main/services/notifications/sns/SNSNotificationServiceImpl.java

```diff
@@ -401,19 +403,44 @@ private void publishNotification(
     ) throws Exception {
         PublishResult result = null;
         try {
-            result = this.client.publish(arn, this.mapper.writeValueAsString(message));
+            result = publishNotification(arn, this.mapper.writeValueAsString(message));
         } catch (Exception exception) {
             log.error("SNS Publish message failed.", exception);
             notificationMetric.counterIncrement(
                 Metrics.CounterSNSNotificationPublishMessageSizeExceeded.getMetricName());
             final SNSMessage<Void> voidMessage = new SNSMessage<>(message.getId(),
                 message.getTimestamp(), message.getRequestId(), message.getType(), message.getName(),
                 null);
-            result = this.client.publish(arn, this.mapper.writeValueAsString(voidMessage));
+            result = publishNotification(arn, this.mapper.writeValueAsString(voidMessage));
         }
         log.info("Successfully published message to topic {} with id {}", arn, result.getMessageId());
         log.debug("Successfully published message {} to topic {} with id {}", message, arn, result.getMessageId());
         notificationMetric.counterIncrement(counterKey);
         notificationMetric.recordTime(message, Metrics.TimerNotificationsPublishDelay.getMetricName());
     }
+
+    private PublishResult publishNotification(final String arn, final String message) {
+        try {
+            return this.client.publish(arn, message);
+        } catch (Exception e) {
+            //
+            // SNS Http client pool once shutdown cannot be recovered. Hence we are shutting down the SNS client
+            // and recreating a new instance.
+            //
+            return Throwables.getCausalChain(e).stream()
+                .filter(ex -> ex instanceof IllegalStateException
+                    && ex.getMessage().contains("Connection pool shut down"))
+                .findFirst()
+                .map(ex -> {
+                    log.warn("SNS HTTP connection pool is down. It will be restarted.");
+                    try {
+                        this.client.shutdown();
+                    } catch (Exception exception) {
+                        log.warn("Failed shutting down SNS client.", exception);
+                    }
```

Comment:
Is this going to be thread-safe? Perhaps we need a lock?

Reply:
Good point. Will need to handle it.